app/(main)/league/all/page.test.tsx

```diff
@@ -1,124 +1,295 @@
-    });
   });
 
-  xtest('should handle form submission to join a league', async () => {
-    mockUseDataStore.mockReturnValueOnce({
-      user: { id: '123', leagues: [] },
+  test('should not display GlobalSpinner after loading data', async () => {
+    mockUseAuthContext.isSignedIn = true;
+
+    mockUseDataStore.mockReturnValue({
+      user: {
+        documentId: '123',
+        email: 'test@test.com',
+        id: '123',
+        leagues: [],
+      },
       allLeagues: [
         {
           leagueId: '123',
           leagueName: 'Test League',
-          logo: 'https://findmylogo.com/logo.png',
+          logo: 'logo.png',
           participants: ['123456', '78'],
-          survivors: ['123456', '78', '9'],
+          survivors: ['123456', '78'],
         },
       ],
     });
+
     mockGetUserLeagues.mockResolvedValueOnce([]);
-    mockGetGameWeek.mockResolvedValueOnce({ week: 1 });
+
+    render(<Leagues />);
+
+    await waitForElementToBeRemoved(() => screen.getByTestId('global-spinner'));
+
+    expect(screen.queryByTestId('global-spinner')).not.toBeInTheDocument();
+  });
+
+  test('should handle form submission to join a league', async () => {
+    mockUseAuthContext.isSignedIn = true;
+
+    const user = {
+      documentId: '123',
+      email: 'test@test.com',
+      id: '123',
+      leagues: ['12'],
+    };
+
+    const league = {
+      leagueId: '123',
+      leagueName: 'Test League',
+      logo: 'logo.png',
+      participants: [],
+      survivors: [],
+    };
+
+    const updateUser = jest.fn();
+
+    mockUseDataStore.mockReturnValue({
+      user,
+      allLeagues: [league],
+      updateUser,
+    });
+
+    mockGetUserLeagues.mockResolvedValueOnce([]);
+    mockGetAllLeagues.mockResolvedValueOnce([league]);
+    mockAddUserToLeague.mockResolvedValue(
+      Promise.resolve({
+        documentId: user.documentId,
+        leagueId: league.leagueId,
+        selectedLeagues: [...(user.leagues ?? []), league.leagueId],
+        participants: [...(league.participants ?? []), user.id],
+        survivors: [...(league.survivors ?? []), user.id],
+      }),
+    );
 
     render(<Leagues />);
 
     await waitFor(() => {
       expect(screen.queryByTestId('global-spinner')).not.toBeInTheDocument();
     });
 
-    const selectElement = screen.getByLabelText(/Select league to join/i);
-    expect(selectElement).toBeInTheDocument();
+    const selectElement = screen.getByTestId('select-available-leagues');
+    fireEvent.change(selectElement, { target: { value: '123' } });
+    fireEvent.click(screen.getByTestId('join-league-button'));
+
+    await waitFor(() => {
+      expect(mockAddUserToLeague).toHaveBeenCalledWith({
+        userDocumentId: user.documentId,
+        selectedLeague: league.leagueId,
+        selectedLeagues: [...(user.leagues ?? []), league.leagueId],
+        participants: [...(league.participants ?? []), user.id],
+        survivors: [...(league.survivors ?? []), user.id],
```

Comment:
you don't need `...(user.leagues ?? [])` for selected, participants or survivors.

all youneed is league.leagueId, user.id, user.id.

```
  userDocumentId: user.documentId,
        selectedLeague: league.leagueId,
        selectedLeagues: [league.leagueId],
        participants: [user.id],
        survivors: [user.id],
        ```

Reply:
The reason why I set up this code like this is to match the code in `league/page.tsx` file. What are your thoughts on the different formats?

Reply:
@vmaineng it's not about different formats. It's about what they are doing.

In page.tsx, the code handles adding a new participant/survivor as follows:

- It spreads the existing participants/survivors (if any).
- It then adds the new participant/survivor to this spread array.
- If there's no existing history of participants/survivors, it starts with an empty array before adding the new user.

The test is verifying that the function is called with the correct data you're passing into it. By having it the same way as page.tsx you are duplicating code.

Reply:
Thank you for explaining it so well, Chris. Updated :)